sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/StringExpressionsSuite.scala

```diff
@@ -1108,6 +1125,238 @@ class StringExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format2))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format3))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+    }
+
+    Seq(
+      (Decimal(12454), "000,000") -> "012,454",
+      (Decimal(12454), "00,0000") -> "01,2454",
+      (Decimal(12454), "000,0000") -> "001,2454",
+      (Decimal(12454), "0000,0000") -> "0001,2454",
+      (Decimal(12454), "00,0000") -> "01,2454",
+      (Decimal(12454), "000,0000") -> "001,2454",
+      (Decimal(12454), "0000,0000") -> "0001,2454",
+      (Decimal(12454367), "000,000,000") -> "012,454,367",
+      (Decimal(12454), "000,000") -> "012,454",
+      (Decimal(12454), "999,999") -> " 12,454",
+      (Decimal(12454), "9,9999") -> "1,2454",
+      (Decimal(12454), "99,9999") -> " 1,2454",
+      (Decimal(12454), "999,9999") -> "  1,2454",
+      (Decimal(12454), "9999,9999") -> "   1,2454",
+      (Decimal(12454367), "999,999,999") -> " 12,454,367",
+      (Decimal(12454), "999,999") -> " 12,454"
+    ).foreach { case ((decimal, format), expected) =>
+      var expr: Expression = ToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+    }
+
+    // Test '$'
+    Seq(
+      (Decimal(78.12), "$99.99") -> "$78.12",
+      (Decimal(78.12), "$00.00") -> "$78.12"
+    ).foreach { case ((decimal, format), expected) =>
+      var expr: Expression = ToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+
+      expr = TryToCharacter(Literal(decimal), Literal(format))
+      assert(expr.checkInputDataTypes() == TypeCheckResult.TypeCheckSuccess)
+      checkEvaluation(expr, expected)
+    }
+
+    // Test 'S'
+    Seq(
+      (Decimal(unscaled = 43100000000L, precision = 38, scale = 10),
+        "9999999999999999.999999999999999S") ->
+        "               4.31              ",
+      (Decimal(0), "9999999999999999.999999999999999S") -> "               0                 ",
+      (Decimal(-454.8), "99G999.9S") -> "   454.8-",
+      (Decimal(-454.8), "00G000.0S") -> "00,454.8-",
+      (Decimal(-454), "S999") -> "-454",
+      (Decimal(-454), "999S") -> "454-",
+      (Decimal(-12454.8), "99G999D9S") -> "12,454.8-",
+      (Decimal(-83028485), "9999999999999999.999999999999999S") ->
+        "        83028485-                "
```

Comment:
can we use smaller numbers? otherwise it's hard to review the test as my eyeball can't count so any whitespaces...

Reply:
BTW can we have a simliar test when `S` is at the beginning?

Reply:
SG, done.